### PR TITLE
dnscrypt-proxy as a dns adblocker

### PR DESCRIPTION
**Pull request by jackivanov**
## Description
As discussed in #1478, dnsmasq can be abandoned in favor of dnscrypt-proxy (it can handle dns ad blocking). In this PR I removed the installation and configuration of dnsmasq and moved dns encryption and dns ad blocking to dnscrypt-proxy.

## Motivation and Context
- removes the `dns_adblocking` role
- renames `dns_encryption` role to `dns`
- moves the ad blocking to dnscrypt-proxy
- renames `local_dns` variable to `dns_adblocking`
- switches back to the ppa, because it updates more frequently and we can keep the same config file for both 18.04 and 19.04

## How Has This Been Tested?
Deployed to lightsail

## Types of changes

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by davidemyers** [comment]
This worked for me deploying to DigitalOcean. I don't normally use DNS adblocking so others should test this since it returns REFUSED instead of NXDOMAIN for blocked domains, perhaps changing browser behavior somehow.

Is this technically a "breaking change" since it might break existing user scripts that deploy by running `ansible-playbook main.yml` directly?

**Comment by TC1977** [comment]
It's a "breaking change" for me, since I use dnsmasq to also serve DNS requests for my home network. But that's okay. I think dnscrypt-proxy will respond to queries from all interfaces by default?

dnscrypt-proxy is being updated pretty frequently, it seems, so I'm adding comments on if we should update versions and default templates.

File: roles/dns/defaults/main.yml [hunk 1]
```diff
@@ -1,6 +1,5 @@
 ---
-algo_local_dns: false
-listen_port: "{% if algo_local_dns %}5353{% else %}53{% endif %}"
+algo_dns_adblocking: false
 # the version used if the latest unavailable (in case of Github API rate limited)
 dnscrypt_proxy_version: 2.0.10
```
Comment (TC1977) [inline roles/dns/defaults/main.yml]: Most recent version is [2.0.25](https://github.com/jedisct1/dnscrypt-proxy/blob/master/ChangeLog), and current version on my Lightsail instance is 2.0.23.

**Comment by jackivanov** [comment]
Here are some benchmark with the default configs provided by Algo. It looks like dnscrypt-proxy with ad blocking has the best results.

- 3 Algo VPN servers (DigitalOcean NYC1)
  - 172.26.114.0 DNScrypt Ad Blocking
  - 172.26.193.98 DNSmasq Encryption
  - 172.20.83.131 DNSmasq
- 1 WireGuard linux client (DigitalOcean NYC1) with all three tunnels up and routes to the DNS IPs set
  - namebench sanity disabled
  - blocked domains added to namebench
  - Google Public DNS 8.8.8.8 added to the testing list

[Response Distribution Chart  for 200ms](http://chart.apis.google.com/chart?cht=lxy&chs=720x415&chxt=x,y&chg=10,20&chxr=0,0,200|1,0,100&chd=t:0,0,1,2,3,3,4,7,12,19,45,65,109|0,0,17,45,56,62,65,70,74,77,81,85,89|0,1,1,2,3,4,4,7,96,474|0,0,34,47,66,82,90,93,97,100|0,1,1,3,4,5,5,6,7,8,21,122|0,0,8,26,46,52,61,79,86,90,94,97|0,1,1,3,4,5,6,7,8,12,25,59,97,246|0,0,6,14,22,60,71,77,82,86,90,93,97,100|0,1,1,5,5,6,8,8,9,14,37,46,56,91,117|0,0,6,31,46,55,58,62,67,73,77,81,85,88,92&chco=ff9900,1a00ff,ff00e6,80ff00,00e6ff&chxt=x,y,x,y&chxl=2:||Duration+in+ms||3:||%25|&chdl=SYS-127.0.0.53|DNScrypt+Ad+Blocking|DNSmasq|DNSmasq+Encryption|Google+Public+DNS)
[Response Distribution Chart Full](http://chart.apis.google.com/chart?cht=lxy&chs=720x415&chxt=x,y&chg=10,20&chxr=0,0,3012|1,0,100&chd=t:0,0,0,0,0,0,0,0,1,1,3,4,7,8,11,27,35|0,0,17,45,56,62,65,70,74,77,81,85,89,92,96,100,100|0,0,0,0,0,0,0,0,6,32|0,0,34,47,66,82,90,93,97,100|0,0,0,0,0,0,0,0,0,1,1,8,57|0,0,8,26,46,52,61,79,86,90,94,97,100|0,0,0,0,0,0,0,0,1,1,2,4,6,16|0,0,6,14,22,60,71,77,82,86,90,93,97,100|0,0,0,0,0,0,1,1,1,1,2,3,4,6,8,13,37,100|0,0,6,31,46,55,58,62,67,73,77,81,85,88,92,96,99,100&chco=ff9900,1a00ff,ff00e6,80ff00,00e6ff&chxt=x,y,x,y&chxl=2:||Duration+in+ms||3:||%25|&chdl=SYS-127.0.0.53|DNScrypt+Ad+Blocking|DNSmasq|DNSmasq+Encryption|Google+Public+DNS)

<details><summary>namebench output</summary>
<p>

```
namebench 1.3.1 - best source (automatic) on 2019-06-14 07:52:52.167674
threads=40/2 queries=250 runs=1 timeout=3.5 health_timeout=3.75 servers=11
------------------------------------------------------------------------------
- Reading Top 2,000 Websites (Alexa): /usr/share/namebench/data/alexa-top-2000-domains.txt (0.7MB)
- Reading Blocked domains: /tmp/blocked.txt (2.0MB)
- Reading Cache Latency Test (100% hit): /usr/share/namebench/data/cache-hit.txt (0.1MB)
- Reading Cache Latency Test (100% miss): /usr/share/namebench/data/cache-miss.txt (0.1MB)
- Reading Cache Latency Test (50% hit, 50% miss): /usr/share/namebench/data/cache-mix.txt (0.1MB)
- Generating tests from Top 2,000 Websites (Alexa) (33575 records, selecting 250 automatic)
- Selecting 250 out of 33542 sanitized records (weighted mode).

- Checking query interception status...
- Checking connection quality: 1/3...3/3
- Congestion level is 0.34X (check duration: 13.77ms)
- Checking latest sanity reference
- Checking nameserver availability (5 threads): 0/5...5/5
- 5 of 5 servers are available (duration: 0:00:00.506733)
- Running initial health checks on 5 servers (5 threads): 0/5...5/5
- 5 of 5 tested name servers are healthy
- Waiting for wildcard cache queries from 5 servers (5 threads): 0/5...5/5
- Waiting 4s for TTL's to decrement.
- Running cache-sharing checks on 5 servers (20 threads): 0/20..........20/20
- Running final health checks on 5 servers (5 threads): 0/5....5/5

Final list of nameservers considered:
------------------------------------------------------------------------------
8.8.8.8         Google Public DNS  8   ms |
127.0.0.53      SYS-127.0.0.53     13  ms |
172.26.114.0    DNScrypt Ad Blocki 24  ms |
172.26.193.98   DNSmasq Encryption 24  ms |
172.20.83.131   DNSmasq            26  ms |

- Sending 250 queries to 5 servers: 0/1250....................................................................................................................................254...................................................................................................................................508....................................................................................................................................766..................................................................................................................................1019.......................................................................................................................1250/1250

Fastest individual response (in milliseconds):
----------------------------------------------
SYS-127.0.0.53   #################### 0.63896
DNSmasq Encrypti #################################### 1.14107
DNScrypt Ad Bloc ######################################## 1.28603
DNSmasq          ################################################### 1.62292
Google Public DN ##################################################### 1.71185

Mean response (in milliseconds):
--------------------------------
DNScrypt Ad Bloc ############# 18.88
DNSmasq Encrypti ################### 26.94
DNSmasq          ######################## 33.73
SYS-127.0.0.53   ############################################ 63.69
Google Public DN ##################################################### 77.50

Response Distribution Chart URL (200ms):
----------------------------------------
http://chart.apis.google.com/chart?cht=lxy&chs=720x415&chxt=x,y&chg=10,20&chxr=0,0,200|1,0,100&chd=t:0,0,1,2,3,3,4,7,12,19,45,65,109|0,0,17,45,56,62,65,70,74,77,81,85,89|0,1,1,2,3,4,4,7,96,474|0,0,34,47,66,82,90,93,97,100|0,1,1,3,4,5,5,6,7,8,21,122|0,0,8,26,46,52,61,79,86,90,94,97|0,1,1,3,4,5,6,7,8,12,25,59,97,246|0,0,6,14,22,60,71,77,82,86,90,93,97,100|0,1,1,5,5,6,8,8,9,14,37,46,56,91,117|0,0,6,31,46,55,58,62,67,73,77,81,85,88,92&chco=ff9900,1a00ff,ff00e6,80ff00,00e6ff&chxt=x,y,x,y&chxl=2:||Duration+in+ms||3:||%25|&chdl=SYS-127.0.0.53|DNScrypt+Ad+Blocking|DNSmasq|DNSmasq+Encryption|Google+Public+DNS

Response Distribution Chart URL (Full):
---------------------------------------
http://chart.apis.google.com/chart?cht=lxy&chs=720x415&chxt=x,y&chg=10,20&chxr=0,0,3012|1,0,100&chd=t:0,0,0,0,0,0,0,0,1,1,3,4,7,8,11,27,35|0,0,17,45,56,62,65,70,74,77,81,85,89,92,96,100,100|0,0,0,0,0,0,0,0,6,32|0,0,34,47,66,82,90,93,97,100|0,0,0,0,0,0,0,0,0,1,1,8,57|0,0,8,26,46,52,61,79,86,90,94,97,100|0,0,0,0,0,0,0,0,1,1,2,4,6,16|0,0,6,14,22,60,71,77,82,86,90,93,97,100|0,0,0,0,0,0,1,1,1,1,2,3,4,6,8,13,37,100|0,0,6,31,46,55,58,62,67,73,77,81,85,88,92,96,99,100&chco=ff9900,1a00ff,ff00e6,80ff00,00e6ff&chxt=x,y,x,y&chxl=2:||Duration+in+ms||3:||%25|&chdl=SYS-127.0.0.53|DNScrypt+Ad+Blocking|DNSmasq|DNSmasq+Encryption|Google+Public+DNS

Recommended configuration (fastest + nearest):
----------------------------------------------
nameserver 172.26.114.0    # DNScrypt Ad Blocking
nameserver 127.0.0.53      # SYS-127.0.0.53
nameserver 172.26.193.98   # DNSmasq Encryption



********************************************************************************
In this test, DNScrypt Ad Blocking is 237.2%: Faster
********************************************************************************


- Saving report to /tmp/namebench_2019-06-14_0754.html
- Saving detailed results to /tmp/namebench_2019-06-14_0754.csv
</p>
</details>

**Comment by davidemyers** [comment]
The change back to the PPA for Disco has changed the response for a blacklisted domain:

`dnscrypt-proxy 2.0.19+ds1-2` (Disco default):
```
root@vpn6:~# host 000free.us 172.20.39.7
Using domain server:
Name: 172.20.39.7
Address: 172.20.39.7#53
Aliases: 

Host 000free.us not found: 5(REFUSED)
```

`dnscrypt-proxy 2.0.25~ppa1~disco` (PPA):
```
root@vpn7:~# host 000free.us 172.28.118.1
Using domain server:
Name: 172.28.118.1
Address: 172.28.118.1#53
Aliases: 

000free.us host information "This query has been locally blocked" "by dnscrypt-proxy"
000free.us host information "This query has been locally blocked" "by dnscrypt-proxy"
000free.us host information "This query has been locally blocked" "by dnscrypt-proxy"
```

The ChangeLog says:

> - New option: `refused_code_in_responses` to return (or not) a
`REFUSED` code on blacklisted queries. This is disabled by default, in
order to work around a bug in Android Pie.

I mention this to again encourage users of DNS adblocking to test this PR to make sure your browsers behave as expected.

**Comment by jackivanov** [comment]
Yes, we need to test both ways of answering before merging this PR.

**Comment by davidemyers** [comment]
@TC1977 I've been using the `forwarding_rules` feature of `dnscrypt-proxy` to forward queries for my home network to my home DNS server (I have my firewall only allow DNS queries from my current AlgoVPN). The only issue I've encountered is that I have to use FQDNs when connected with WireGuard because there's no way to specify a default search domain.

**Comment by jackivanov** [comment]
I have some delays surfing sites in Safari, but not in Chrome with `refused_code_in_responses = false`
(eg: this url https://b.cdnst.net/javascript/amazon.js)

File: input.yml [hunk 2]
```diff
@@ -99,10 +99,10 @@
       - name: DNS adblocking prompt
         pause:
           prompt: |
-            Do you want to install an ad blocking DNS resolver on this VPN server?
+            Do you want to configure the DNS ad blocking on this VPN server?
```
Comment (TC1977) [inline input.yml]: I would suggest "do you want to enable DNS ad blocking" instead. "Configure" can be ambiguous.

**Comment by TC1977** [comment]
For what it's worth, I've installed this branch with default `refused_code_in_responses` and have no problems browsing in Safari, iOS, Wireguard, IPv4 and IPv6. No ads either.

**Pull request merged**